client/api_client.go

```diff
@@ -0,0 +1,494 @@
+package client
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/hex"
+	"encoding/xml"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	hashlib "github.com/bnb-chain/greenfield-common/go/hash"
+	httplib "github.com/bnb-chain/greenfield-common/go/http"
+	"github.com/bnb-chain/greenfield-go-sdk/pkg/utils"
+	"github.com/bnb-chain/greenfield-go-sdk/types"
+	sdkclient "github.com/bnb-chain/greenfield/sdk/client"
+	gnfdSdkTypes "github.com/bnb-chain/greenfield/sdk/types"
+	permTypes "github.com/bnb-chain/greenfield/x/permission/types"
+	storageTypes "github.com/bnb-chain/greenfield/x/storage/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+type Client interface {
+	Basic
+	Bucket
+	Object
+	Group
+	Challenge
+	Account
+	SP
+}
+
+// client represents a Greenfield SDK client that can interact with the blockchain
+// using the REST API, gRPC, or WebSocket endpoints.
+type client struct {
+	// The chain client is used to interact with the blockchain
+	chainClient *sdkclient.GreenfieldClient
+	// The HTTP client is used to send HTTP requests to the greenfield blockchain and sp
+	httpClient *http.Client
+	// Service provider endpoints
+	spEndpoints map[string]*url.URL
+	// The default account to use when sending transactions.
+	defaultAccount *types.Account
+	// Whether the connection to the blockchain node is secure (HTTPS) or not (HTTP).
+	secure bool
+	// Host is the target sp server hostname，it is the host info in the request which sent to SP
+	host string
+	// The user agent info
+	userAgent string
+}
+
+// Option is a configuration struct used to provide optional parameters to the client constructor.
+type Option struct {
+	// GrpcDialOption is the list of gRPC dial options used to configure the connection to the blockchain node.
+	GrpcDialOption grpc.DialOption
+	// Secure is a flag that specifies whether the client should use HTTPS or not.
+	Secure bool
+	// Transport is the HTTP transport used to send requests to the storage provider endpoint.
+	Transport http.RoundTripper
+	// Host is the target sp server hostname
+	Host string
+}
+
+// New - instantiate greenfield chain with chain info, account info and options.
+// The grpcAddress indicate the grpc address of greenfield chain.
+// The account indicate the account used for signing transactions.
+// The rpcAddress indicate the rpc address of the tendermint client.
+func New(chainID string, grpcAddress string, account *types.Account, option Option) (Client, error) {
+	if grpcAddress == "" || chainID == "" {
+		return nil, errors.New("fail to get grpcAddress and chainID to construct client")
+	}
+
+	if account == nil {
+		return nil, errors.New("fail to get account info")
```

Comment:
And we should add another API to set the default account dynamic

Reply:
Sure. I put the account to options, and add a `MustGetDefaultAccount` function to client. 

Therefore, if any interface needs to use the account to send request or tx, `MustGetDefaultAccount` must be called, and if the default account does not exist, it will panic

Reply:
We now provide an API called `SetDefaultAccount` to set the default account dynamic.